### PR TITLE
[+] Search - Do not return fields that are hidden from the UI except if a smart field is displayed.

### PR DESCRIPTION
**Pull request by slimee**
…that are hidden from the UI

Depending on which field is hidden on the UI, field selection is passed to Lumber API (ressources-getter).
Before this PR, mongoose liana returns always all fields from search query.
After this PR, mongoose liana return only fields that are selected, **when there is no smart field in the selected fields selection**

## Pull Request checklist:

- [x] Write an explicit title for the Pull Request
- [x] Write changes made in the CHANGELOG.md
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

**Comment by slimee** [comment]
[Mongo liana - UI Hidden fields are returned from Liana.](https://trello.com/c/8ElRGtF1/3916-p2-mongo-liana-ui-hidden-fields-are-returned-from-liana)

File: test/tests/services/projection-builder.test.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        ],
+      };
+      const projectionBuilder = new ProjectionBuilder(schemaWithSmartFields);
+      const expectedProjection = null;
+      const actualProjection = projectionBuilder.getProjection(['fullname']);
+
+      expect(actualProjection).toStrictEqual(expectedProjection);
+    });
+  });
+});
```
Comment (rap2hpoutre) [inline test/tests/services/projection-builder.test.js]: Maybe add one test like `getProjection with empty request` but not sure it's an actual case 😅 .

File: test/tests/services/resources-getter.test.js [hunk 2]
```diff
@@ -273,4 +310,44 @@ describe('service > resources-getter', () => {
...
+          searchExtended: '0',
+          timezone: '+01:00',
+        };
+
+        const result = await new ResourcesGetter(FilmModel, options, parameters).perform();
+        expect(result[0]).toHaveLength(3);
+        const titles = result[0].filter((film) => !!film.title);
+        expect(titles).toHaveLength(3);
+        const durations = result[0].filter((film) => !!film.duration);
+        expect(durations).toHaveLength(3);
```
Comment (rap2hpoutre) [inline test/tests/services/resources-getter.test.js]: Opinion mode, sorry, please ignore me! (I removed the `[0]` because ... uh ... I feel it noisy and I added a few new lines)

```javascript
const [result] = await new ResourcesGetter(FilmModel, options, parameters).perform();
expect(result).toHaveLength(3);

const titles = result.filter((film) => !!film.title);
expect(titles).toHaveLength(3);

const durations = result.filter((film) => !!film.duration);
expect(durations).toHaveLength(3);
```

File: test/tests/services/resources-getter.test.js [hunk 3]
```diff
@@ -273,4 +310,44 @@ describe('service > resources-getter', () => {
...
+          searchExtended: '0',
+          timezone: '+01:00',
+        };
+
+        const result = await new ResourcesGetter(FilmModel, options, parameters).perform();
+        expect(result[0]).toHaveLength(3);
+        const titles = result[0].filter((film) => !!film.title);
+        expect(titles).toHaveLength(3);
+        const durations = result[0].filter((film) => !!film.duration);
+        expect(durations).toHaveLength(0);
```
Comment (rap2hpoutre) [inline test/tests/services/resources-getter.test.js]: Opini... Yes. Same comment as above, trivial too

File: src/services/projection-builder.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      return { $project: fieldsObject };
+    }
+    return null;
+  }
+
+  // NOTICE: Perform the intersection between schema and request smart fields.
+  findRequestSmartField(requestFieldsNames) {
+    if (this.schemaSmartFields && requestFieldsNames) {
+      return this.schemaSmartFields
+        .filter((fieldName) => requestFieldsNames.indexOf(fieldName) > -1);
```
[the remark this answers is not in the document]
Comment (slimee) [reply src/services/projection-builder.js]: yes thanks!

**Comment by larcin** [comment]
[Misc] It seems that you can steel filter on hidden fields. It's probably out of scope but isn't it a bit weird ?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
 # Change Log
 
 ## [Unreleased]
+### Changed
+- Search - In a request with no smart fields, do not return fields that are hidden from the UI.
```
Comment (arnaudbesnier) [inline CHANGELOG.md]: "Search" is a bit misleading. This change is not related to the "search" feature.
It is about performance optimization.

**Comment by arnaudbesnier** [comment]
> [Misc] It seems that you can still filter on hidden fields. It's probably out of scope but isn't it a bit weird ?

@larcin, you need to be able to "filter" on hidden fields. That's how it works today.
If you have a segment created with conditions on hidden columns, you expect that the API will return the filtered values, right?

**Comment by larcin** [comment]
> > [Misc] It seems that you can still filter on hidden fields. It's probably out of scope but isn't it a bit weird ?
> 
> @larcin, you need to be able to "filter" on hidden fields. That's how it works today.
> If you have a segment created with conditions on hidden columns, you expect that the API will return the filtered values, right?

Ok, I got it now, thanks
